File: samples/python/a2a_skeleton/search_agent/agent/agent_executor.py
from a2a.server.agent_execution import AgentExecutor, RequestContext
from a2a.server.events.event_queue import EventQueue
from a2a.types import TaskArtifactUpdateEvent, TaskStatusUpdateEvent, TaskStatus, TaskState
from a2a.utils import new_agent_text_message, new_task, new_text_artifact
from a2a.utils.errors import ServerError
from a2a.types import TaskState, TextPart, UnsupportedOperationError, Message

from agent.agent import SearchAgent
from a2a.types import AgentCard
import logging
logger = logging.getLogger("search_agent.agent_executor")

class SearchAgentExecutor(AgentExecutor):
    def __init__(self):
        self.agent = SearchAgent()

    async def execute(
        self,
        context: RequestContext,
        event_queue: EventQueue,
    ) -> None:
        metadata = context._params.message.metadata
        session_id = metadata.get("session_id", None)
        user_id = metadata.get("user_id", None)
        query = context.get_user_input()
        task = context.current_task
        try:
            if not task :
                task = new_task(context.message)
                await event_queue.enqueue_event(task)
            logger.info(f"invoke 진입전 ")
            async for event in self.agent.invoke(query, session_id, task.id, user_id):
                event_dict = event.dict()
                text = event_dict['content']['parts'][0]['text']
                is_final = event.is_final_response()
                if is_final:
                    logger.info(f"event: {event}")
                    await event_queue.enqueue_event(
                        TaskArtifactUpdateEvent(
                            taskId=task.id,
                            contextId=task.contextId,
                            artifact=new_text_artifact(
                                name='search_result',
                                description='검색 결과',
                                text=text,
                            ),
                            append=False,
                            lastChunk=True,
                        )
                    )
                    await event_queue.enqueue_event(
                        TaskStatusUpdateEvent(
                            taskId=task.id,
                            contextId=task.contextId,
                            status=TaskStatus(state=TaskState.completed),
                            final=True
                        )
                    )
                else :
                    await event_queue.enqueue_event(
                        TaskStatusUpdateEvent(
                            taskId=task.id,
                            contextId=task.contextId,
                            status=TaskStatus(
                                state=TaskState.working,
                                message=new_agent_text_message(
                                    text,
                                    task.id,
                                    task.contextId,
                                ),
                            ),
                            final=False,
                        )
                    )
        except Exception as e:
            logger.error(f"[EXECUTE] 예외 발생: {e}", exc_info=True)
            raise

    async def cancel(self, context: RequestContext, event_queue: EventQueue):
        """Cancels an ongoing operation."""
        raise ServerError(error=UnsupportedOperationError())      

File: samples/js/agents/sign-and-verify-agent-card/agent_executor.ts
import { v4 as uuidv4 } from 'uuid';

import {
  Artifact,
  Role,
  Task,
  TaskArtifactUpdateEvent,
  TaskState,
  TaskStatusUpdateEvent,
} from '@a2a-js/sdk';
import {
  AgentEvent,
  AgentExecutor,
  ExecutionEventBus,
  RequestContext,
  UnsupportedOperationError,
} from '@a2a-js/sdk/server';

// --8<-- [start:SignedAgentExecutor]
export class SignedAgentExecutor implements AgentExecutor {
  /** Test AgentProxy Implementation. */

  async execute(
    requestContext: RequestContext,
    eventBus: ExecutionEventBus
  ): Promise<void> {
    /** Execute the agent and manage task lifecycle events. */
    const userMessage = requestContext.userMessage;
    const existingTask = requestContext.task;
    const taskId = requestContext.taskId;
    const contextId = requestContext.contextId;

    // 1. Collect a task from request context
    if (!existingTask) {
      const initialTask: Task = {
        id: taskId,
        contextId: contextId,
        status: {
          state: TaskState.TASK_STATE_SUBMITTED,
          timestamp: new Date().toISOString(),
          message: undefined,
        },
        artifacts: [],
        history: [userMessage],
        metadata: userMessage.metadata,
      };
      eventBus.publish(AgentEvent.task(initialTask));
    }

    // 2. Update task status to working
    const workingStatusUpdate: TaskStatusUpdateEvent = {
      taskId: taskId,
      contextId: contextId,
      status: {
        state: TaskState.TASK_STATE_WORKING,
        message: {
          role: Role.ROLE_AGENT,
          messageId: uuidv4(),
          parts: [
            {
              content: { $case: 'text', value: 'Processing verification request...' },
              metadata: undefined,
              filename: '',
              mediaType: 'text/plain',
            },
          ],
          taskId: taskId,
          contextId: contextId,
          extensions: [],
          metadata: {},
          referenceTaskIds: [],
        },
        timestamp: new Date().toISOString(),
      },
      metadata: {},
    };
    eventBus.publish(AgentEvent.statusUpdate(workingStatusUpdate));

    // 3. Collect user request from request content and generate content
    const textPart = userMessage.parts.find(
      (p) => p.content?.$case === 'text'
    );
    const query =
      textPart?.content?.$case === 'text'
        ? textPart.content.value.trim()
        : '';

    const result = query ? `Verify me! (${query})` : 'Verify me!';

    // 4. Add generated response as an artifact
    const artifactId = uuidv4();
    const resultArtifact: Artifact = {
      artifactId: artifactId,
      name: 'Result',
      description: 'Verification result artifact.',
      parts: [
        {
          content: { $case: 'text', value: result },
          metadata: undefined,
          filename: '',
          mediaType: 'text/plain',
        },
      ],
      metadata: undefined,
      extensions: [],
    };

    const artifactUpdate: TaskArtifactUpdateEvent = {
      taskId: taskId,
      contextId: contextId,
      artifact: resultArtifact,
      lastChunk: true,
      append: false,
      metadata: undefined,
    };
    eventBus.publish(AgentEvent.artifactUpdate(artifactUpdate));

    // 5. Update task status to completed
    const finalUpdate: TaskStatusUpdateEvent = {
      taskId: taskId,
      contextId: contextId,
      status: {
        state: TaskState.TASK_STATE_COMPLETED,
        message: {
          role: Role.ROLE_AGENT,
          messageId: uuidv4(),
          parts: [
            {
              content: { $case: 'text', value: 'Request is completed!' },
              metadata: undefined,
              filename: '',
              mediaType: 'text/plain',
            },
          ],
          taskId: taskId,
          contextId: contextId,
          extensions: [],
          metadata: {},
          referenceTaskIds: [],
        },
        timestamp: new Date().toISOString(),
      },
      metadata: undefined,
    };
    eventBus.publish(AgentEvent.statusUpdate(finalUpdate));
  }

  public cancelTask = async (
    _taskId: string,
    _eventBus: ExecutionEventBus
  ): Promise<void> => {
    /** Raise exception as cancel is not supported. */
    throw new UnsupportedOperationError('Cancel is not supported.');
  };
}
// --8<-- [end:SignedAgentExecutor]
